File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-02-preparacao-datasets/dataset-preparation-pipeline-go/cleaning/minhashlsh.go
package cleaning

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

const (
	LimiarDuplicata = 0.55
	MinhashK        = 32
	MinhashSemente  = 42
	LshBandas       = 8
	LshLinhas       = 4
	primoMersenne   = 2147483647 // 2^31 - 1
)

var espacos = regexp.MustCompile(`\s+`)

func NormalizarTexto(texto string) string {
	return strings.TrimSpace(espacos.ReplaceAllString(strings.ToLower(texto), " "))
}

// Shingles gera os n-gramas de palavras (shingles) do texto normalizado.
func Shingles(texto string, n int) map[string]struct{} {
	palavras := strings.Split(NormalizarTexto(texto), " ")
	conjunto := map[string]struct{}{}
	for i := 0; i <= len(palavras)-n; i++ {
		conjunto[strings.Join(palavras[i:i+n], " ")] = struct{}{}
	}
	return conjunto
}

func SimilaridadeJaccardExata(a, b string, n int) float64 {
	sa := Shingles(a, n)
	sb := Shingles(b, n)
	if len(sa) == 0 || len(sb) == 0 {
		return 0
	}
	intersecao := 0
	for s := range sa {
		if _, ok := sb[s]; ok {
			intersecao++
		}
	}
	uniao := len(sa) + len(sb) - intersecao
	if uniao == 0 {
		return 0
	}
	return float64(intersecao) / float64(uniao)
}

// hashString: djb2, hash de string deterministico, 32 bits sem sinal.
func hashString(s string) uint32 {
	var h uint32 = 5381
	for i := 0; i < len(s); i++ {
		h = (h*33 + uint32(s[i])) & 0xffffffff
	}
	return h
}

type CoeficienteHash struct {
	A, B int64
}

// GerarCoeficientesHash: LCG deterministico (mesma semente sempre produz a
// mesma familia de funcoes hash - necessario pra reprodutibilidade).
func GerarCoeficientesHash(k int, semente int64) []CoeficienteHash {
	estado := semente & 0xffffffff
	proximo := func() int64 {
		estado = (estado*1103515245 + 12345) & 0xffffffff
		return estado
	}
	coeficientes := make([]CoeficienteHash, 0, k)
	for i := 0; i < k; i++ {
		a := proximo()%2000000000 + 1
		b := proximo() % 2000000000
		coeficientes = append(coeficientes, CoeficienteHash{A: a, B: b})
	}
	return coeficientes
}

func hashUniversal(x, a, b int64) int64 {
	return (a*x + b) % primoMersenne
}

// AssinaturaMinHash: k valores minimos, um por funcao hash, sobre o conjunto de shingles.
func AssinaturaMinHash(shingleSet map[string]struct{}, coeficientes []CoeficienteHash) []int64 {
	baseHashes := make([]int64, 0, len(shingleSet))
	for s := range shingleSet {
		baseHashes = append(baseHashes, int64(hashString(s)))
	}
	assinatura := make([]int64, len(coeficientes))
	for i, c := range coeficientes {
		minimo := int64(1) << 62
		for _, x := range baseHashes {
			h := hashUniversal(x, c.A, c.B)
			if h < minimo {
				minimo = h
			}
		}
		assinatura[i] = minimo
	}
	return assinatura
}

func SimilaridadeMinHashEstimada(a, b []int64) float64 {
	iguais := 0
	for i := range a {
		if a[i] == b[i] {
			iguais++
		}
	}
	return float64(iguais) / float64(len(a))
}

// BandingLSH: divide a assinatura de k valores em b bandas de r valores cada
// (k = b*r). Dois exemplos viram "candidatos" se colidirem em pelo menos uma banda.
func BandingLSH(assinaturas [][]int64, b, r int) map[string]struct{} {
	baldes := map[string][]int{}
	candidatos := map[string]struct{}{}
	for idx, assinatura := range assinaturas {
		for banda := 0; banda < b; banda++ {
			partes := make([]string, r)
			for j := 0; j < r; j++ {
				partes[j] = strconv.FormatInt(assinatura[banda*r+j], 10)
			}
			chave := fmt.Sprintf("%d:%s", banda, strings.Join(partes, ","))
			for _, outroIdx := range baldes[chave] {
				menor, maior := outroIdx, idx
				if idx < outroIdx {
					menor, maior = idx, outroIdx
				}
				candidatos[fmt.Sprintf("%d-%d", menor, maior)] = struct{}{}
			}
			baldes[chave] = append(baldes[chave], idx)
		}
	}
	return candidatos
}

type ParDuplicata struct {
	I, J         int
	Similaridade float64
}

type ResultadoPorCaso struct {
	ItensNoCaso           int
	ParesForcaBruta       int64
	CandidatosLSH         int
	DuplicatasConfirmadas int
	ReducaoPercentual     float64
}

type ResultadoDedup struct {
	ParesDuplicata       []ParDuplicata
	ResultadosPorCaso    map[string]ResultadoPorCaso
	TotalParesForcaBruta int64
	TotalCandidatosLSH   int64
}

func EncontrarQuaseDuplicatasMinHashLSH(exemplos []Exemplo) ResultadoDedup {
	coeficientes := GerarCoeficientesHash(MinhashK, MinhashSemente)
	resultadosPorCaso := map[string]ResultadoPorCaso{}
	var totalParesForcaBruta, totalCandidatosLSH int64
	var paresDuplicata []ParDuplicata

	for _, caso := range []string{"amplitude-auto", "amplitude-saude-empresarial"} {
		var indicesGlobais []int
		var itens []Exemplo
		for i, e := range exemplos {
			if e.Metadata.Caso == caso {
				indicesGlobais = append(indicesGlobais, i)
				itens = append(itens, e)
			}
		}
		assinaturas := make([][]int64, len(itens))
		for i, e := range itens {
			assinaturas[i] = AssinaturaMinHash(Shingles(e.Entrada, 5), coeficientes)
		}
		candidatosLocais := BandingLSH(assinaturas, LshBandas, LshLinhas)
		paresForcaBruta := int64(len(itens)) * int64(len(itens)-1) / 2

		confirmados := 0
		for chave := range candidatosLocais {
			var li, lj int
			fmt.Sscanf(chave, "%d-%d", &li, &lj)
			simExata := SimilaridadeJaccardExata(itens[li].Entrada, itens[lj].Entrada, 5)
			if simExata >= LimiarDuplicata {
				confirmados++
				paresDuplicata = append(paresDuplicata, ParDuplicata{I: indicesGlobais[li], J: indicesGlobais[lj], Similaridade: simExata})
			}
		}

		reducaoPercentual := 0.0
		if paresForcaBruta != 0 {
			reducaoPercentual = 100 * (1 - float64(len(candidatosLocais))/float64(paresForcaBruta))
		}
		resultadosPorCaso[caso] = ResultadoPorCaso{
			ItensNoCaso: len(itens), ParesForcaBruta: paresForcaBruta, CandidatosLSH: len(candidatosLocais),
			DuplicatasConfirmadas: confirmados, ReducaoPercentual: reducaoPercentual,
		}
		totalParesForcaBruta += paresForcaBruta
		totalCandidatosLSH += int64(len(candidatosLocais))
	}

	return ResultadoDedup{
		ParesDuplicata: paresDuplicata, ResultadosPorCaso: resultadosPorCaso,
		TotalParesForcaBruta: totalParesForcaBruta, TotalCandidatosLSH: totalCandidatosLSH,
	}
}

type ResultadoRemocao struct {
	Mantidos             []Exemplo
	ParesDuplicata       []ParDuplicata
	ResultadosPorCaso    map[string]ResultadoPorCaso
	TotalParesForcaBruta int64
	TotalCandidatosLSH   int64
	Removidos            int
}

// RemoverQuaseDuplicatas mantem a primeira ocorrencia de cada par de
// quase-duplicata e remove a segunda.
func RemoverQuaseDuplicatas(exemplos []Exemplo) ResultadoRemocao {
	dedup := EncontrarQuaseDuplicatasMinHashLSH(exemplos)
	remover := map[int]struct{}{}
	for _, p := range dedup.ParesDuplicata {
		remover[p.J] = struct{}{}
	}
	var mantidos []Exemplo
	for i, e := range exemplos {
		if _, ok := remover[i]; !ok {
			mantidos = append(mantidos, e)
		}
	}
	return ResultadoRemocao{
		Mantidos: mantidos, ParesDuplicata: dedup.ParesDuplicata, ResultadosPorCaso: dedup.ResultadosPorCaso,
		TotalParesForcaBruta: dedup.TotalParesForcaBruta, TotalCandidatosLSH: dedup.TotalCandidatosLSH,
		Removidos: len(remover),
	}
}

File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-02-preparacao-datasets/dataset-preparation-pipeline-java/src/main/java/com/datasetprep/cleaning/MinHashLsh.java
package com.datasetprep.cleaning;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Porte da secao 2 de dataset-cleaning-balancing-tool.js: deduplicacao via
 * MinHash (Broder 1997) + LSH banding para geracao de candidatos, com
 * refinamento por similaridade de Jaccard exata so sobre os candidatos.
 * Mesma familia de tecnica do NEARDUP de Lee et al. 2021/2022 (arXiv 2107.06499).
 */
public final class MinHashLsh {

    private MinHashLsh() {
    }

    public static final double LIMIAR_DUPLICATA = 0.55;
    public static final int MINHASH_K = 32;
    public static final int MINHASH_SEMENTE = 42;
    public static final int LSH_BANDAS = 8;
    public static final int LSH_LINHAS = 4;

    private static final long PRIMO_MERSENNE = 2147483647L; // 2^31 - 1

    public static String normalizarTexto(String texto) {
        return texto.toLowerCase().replaceAll("\\s+", " ").trim();
    }

    public static Set<String> shingles(String texto, int n) {
        String[] palavras = normalizarTexto(texto).split(" ");
        Set<String> conjunto = new LinkedHashSet<>();
        for (int i = 0; i <= palavras.length - n; i++) {
            StringBuilder sb = new StringBuilder();
            for (int j = i; j < i + n; j++) {
                if (j > i) sb.append(' ');
                sb.append(palavras[j]);
            }
            conjunto.add(sb.toString());
        }
        return conjunto;
    }

    public static double similaridadeJaccardExata(String a, String b, int n) {
        Set<String> sa = shingles(a, n);
        Set<String> sb = shingles(b, n);
        if (sa.isEmpty() || sb.isEmpty()) return 0;
        long intersecao = sa.stream().filter(sb::contains).count();
        long uniao = sa.size() + sb.size() - intersecao;
        return uniao == 0 ? 0 : (double) intersecao / uniao;
    }

    public static double similaridadeJaccardExata(String a, String b) {
        return similaridadeJaccardExata(a, b, 5);
    }

    /** djb2, hash de string deterministico, 32 bits sem sinal. */
    public static long hashString(String s) {
        long h = 5381;
        for (int i = 0; i < s.length(); i++) {
            h = ((h * 33) + s.charAt(i)) & 0xffffffffL;
        }
        return h;
    }

    public record CoeficienteHash(long a, long b) {
    }

    /** LCG deterministico: mesma semente sempre produz a mesma familia de funcoes hash. */
    public static List<CoeficienteHash> gerarCoeficientesHash(int k, long semente) {
        long[] estado = {semente & 0xffffffffL};
        List<CoeficienteHash> coeficientes = new ArrayList<>();
        for (int i = 0; i < k; i++) {
            estado[0] = ((estado[0] * 1103515245L) + 12345L) & 0xffffffffL;
            long a = (estado[0] % 2000000000L) + 1;
            estado[0] = ((estado[0] * 1103515245L) + 12345L) & 0xffffffffL;
            long b = estado[0] % 2000000000L;
            coeficientes.add(new CoeficienteHash(a, b));
        }
        return coeficientes;
    }

    public static long hashUniversal(long x, long a, long b) {
        return ((a * x) + b) % PRIMO_MERSENNE;
    }

    /** Assinatura MinHash: k valores minimos, um por funcao hash, sobre o conjunto de shingles. */
    public static long[] assinaturaMinHash(Set<String> shingleSet, List<CoeficienteHash> coeficientes) {
        long[] baseHashes = shingleSet.stream().mapToLong(MinHashLsh::hashString).toArray();
        long[] assinatura = new long[coeficientes.size()];
        for (int i = 0; i < coeficientes.size(); i++) {
            CoeficienteHash c = coeficientes.get(i);
            long minimo = Long.MAX_VALUE;
            for (long x : baseHashes) {
                long h = hashUniversal(x, c.a(), c.b());
                if (h < minimo) minimo = h;
            }
            assinatura[i] = minimo;
        }
        return assinatura;
    }

    public static double similaridadeMinHashEstimada(long[] a, long[] b) {
        int iguais = 0;
        for (int i = 0; i < a.length; i++) if (a[i] == b[i]) iguais++;
        return (double) iguais / a.length;
    }

    /**
     * LSH banding: divide a assinatura de k valores em b bandas de r valores cada
     * (k = b*r). Dois exemplos viram "candidatos" se colidirem em pelo menos uma banda.
     */
    public static Set<String> bandingLSH(List<long[]> assinaturas, int b, int r) {
        Map<String, List<Integer>> baldes = new HashMap<>();
        Set<String> candidatos = new LinkedHashSet<>();
        for (int idx = 0; idx < assinaturas.size(); idx++) {
            long[] assinatura = assinaturas.get(idx);
            for (int banda = 0; banda < b; banda++) {
                StringBuilder fatia = new StringBuilder();
                for (int j = banda * r; j < banda * r + r; j++) {
                    if (j > banda * r) fatia.append(',');
                    fatia.append(assinatura[j]);
                }
                String chave = banda + ":" + fatia;
                List<Integer> balde = baldes.computeIfAbsent(chave, k -> new ArrayList<>());
                for (int outroIdx : balde) {
                    int menor = Math.min(outroIdx, idx);
                    int maior = Math.max(outroIdx, idx);
                    candidatos.add(menor + "-" + maior);
                }
                balde.add(idx);
            }
        }
        return candidatos;
    }

    public record ParDuplicata(int i, int j, double similaridade) {
    }

    public record ResultadoPorCaso(
            int itensNoCaso, long paresForcaBruta, int candidatosLSH, int duplicatasConfirmadas, double reducaoPercentual
    ) {
    }

    public record ResultadoDedup(
            List<ParDuplicata> paresDuplicata,
            Map<String, ResultadoPorCaso> resultadosPorCaso,
            long totalParesForcaBruta,
            long totalCandidatosLSH
    ) {
    }

    public static ResultadoDedup encontrarQuaseDuplicatasMinHashLSH(List<Exemplo> exemplos) {
        List<CoeficienteHash> coeficientes = gerarCoeficientesHash(MINHASH_K, MINHASH_SEMENTE);
        Map<String, ResultadoPorCaso> resultadosPorCaso = new HashMap<>();
        long totalParesForcaBruta = 0;
        long totalCandidatosLSH = 0;
        List<ParDuplicata> paresDuplicata = new ArrayList<>();

        for (String caso : List.of("amplitude-auto", "amplitude-saude-empresarial")) {
            List<Integer> indicesGlobais = new ArrayList<>();
            List<Exemplo> itens = new ArrayList<>();
            for (int i = 0; i < exemplos.size(); i++) {
                if (exemplos.get(i).metadata().caso().equals(caso)) {
                    indicesGlobais.add(i);
                    itens.add(exemplos.get(i));
                }
            }
            List<long[]> assinaturas = new ArrayList<>();
            for (Exemplo e : itens) {
                assinaturas.add(assinaturaMinHash(shingles(e.entrada(), 5), coeficientes));
            }
            Set<String> candidatosLocais = bandingLSH(assinaturas, LSH_BANDAS, LSH_LINHAS);
            long paresForcaBruta = (long) itens.size() * (itens.size() - 1) / 2;

            int confirmados = 0;
            for (String chave : candidatosLocais) {
                String[] partes = chave.split("-");
                int li = Integer.parseInt(partes[0]);
                int lj = Integer.parseInt(partes[1]);
                double simExata = similaridadeJaccardExata(itens.get(li).entrada(), itens.get(lj).entrada());
                if (simExata >= LIMIAR_DUPLICATA) {
                    confirmados++;
                    paresDuplicata.add(new ParDuplicata(indicesGlobais.get(li), indicesGlobais.get(lj), simExata));
                }
            }

            double reducaoPercentual = paresForcaBruta == 0 ? 0 : 100 * (1 - (double) candidatosLocais.size() / paresForcaBruta);
            resultadosPorCaso.put(caso, new ResultadoPorCaso(itens.size(), paresForcaBruta, candidatosLocais.size(), confirmados, reducaoPercentual));
            totalParesForcaBruta += paresForcaBruta;
            totalCandidatosLSH += candidatosLocais.size();
        }

        return new ResultadoDedup(paresDuplicata, resultadosPorCaso, totalParesForcaBruta, totalCandidatosLSH);
    }

    public record ResultadoRemocao(
            List<Exemplo> mantidos,
            List<ParDuplicata> paresDuplicata,
            Map<String, ResultadoPorCaso> resultadosPorCaso,
            long totalParesForcaBruta,
            long totalCandidatosLSH,
            int removidos
    ) {
    }

    /** Mantem a primeira ocorrencia de cada par de quase-duplicata, remove a segunda. */
    public static ResultadoRemocao removerQuaseDuplicatas(List<Exemplo> exemplos) {
        ResultadoDedup dedup = encontrarQuaseDuplicatasMinHashLSH(exemplos);
        Set<Integer> remover = new LinkedHashSet<>();
        for (ParDuplicata p : dedup.paresDuplicata()) remover.add(p.j());

        List<Exemplo> mantidos = new ArrayList<>();
        for (int i = 0; i < exemplos.size(); i++) {
            if (!remover.contains(i)) mantidos.add(exemplos.get(i));
        }
        return new ResultadoRemocao(mantidos, dedup.paresDuplicata(), dedup.resultadosPorCaso(),
                dedup.totalParesForcaBruta(), dedup.totalCandidatosLSH(), remover.size());
    }
}
